hw_decorator.py

```diff
@@ -0,0 +1,124 @@
+from functools import wraps
+from datetime import datetime
+
+
+# ЗАДАЧА-1
+# Написать свой декоратор который будет проверять остаток от деления числа 100 на результат работы функции ниже.
+# Если остаток от деления = 0, вывести сообщение "We are OK!», иначе «Bad news guys, we got {}» остаток от деления.
+def remainder_of_division(f):
+    @wraps(f)
+    def wrapper():
+        divider = f()
+        rez = 100 % int(divider)
+        if rez:
+            return f"Bad news guys, we got {rez}"
+        return "We are OK!"
+
+    return wrapper
+
+@remainder_of_division
+def func():
+    return input("введите число: ")
+
+print(func())
+
+# ЗАДАЧА-2
+# Написать декоратор который будет выполнять предпроверку типа аргумента который передается в вашу функцию.
+# Если это int, тогда выполнить функцию и вывести результат, если это str(),
+# тогда зарейзить ошибку ValueError (raise ValueError(“string type is not supported”))
+
+def check_type(func):
+    @wraps(func)
+    def wrapper(arg):
+        if isinstance(arg, int):
+            return func(arg)
+        elif isinstance(arg, str):
+            return 'ValueError: string type is not supported'
+        return 'ValueError: unknown type'
+
+    return wrapper
+
+
+@check_type
+def func1(arg):
+    return arg
+
+
+
+print(func1("sfffa"))
+print(func1(10.0))
+
+# ЗАДАЧА-3
+# Написать декоратор который будет кешировать значения аргументов и результаты работы вашей функции и записывать
+# его в переменную cache. Если аргумента нет в переменной cache и функция выполняется, вывести сообщение
+# «Function executed with counter = {}, function result = {}» и количество раз сколько эта функция выполнялась.
+# Если значение берется из переменной cache, вывести сообщение «Used cache with counter = {}» и
+# количество раз обращений в cache.
+
+def memoize(func):
+    cache = {}
+    count_func_executed = 0
+    count_used_cache = 0
+
+    @wraps(func)
+    def wrapper(*args):
+        if args in cache:
+            nonlocal count_used_cache
+            count_used_cache += 1
+            print(f'Used cache with counter = {count_used_cache}. Value = {cache[args]}')
+        else:
+            nonlocal count_func_executed
```

Comment:
wrapper.count_used_cache and wrapper.count_func_executed are more preferable than nonlocal variables as they can be re-used